R/model-wrappers.R

```diff
@@ -290,3 +290,173 @@ forecastHybrid_model <- function(y = NULL, samples = NULL,
+#' models <- list(
+#'   "ARIMA" = function(...){EpiSoon::fable_model(model = fable::ARIMA(y), ...)},
+#'   "ETS" = function(...){EpiSoon::fable_model(model = fable::ETS(y), ...)},
+#'   "Drift" = function(...){EpiSoon::fable_model(model = fable::RW(y ~ drift()), ...)}
+#' )
+#'
+#' # make forecst on its own
+#' forecast <- stackr_mixture_model(y = EpiSoon::example_obs_rts[1:10, ]$rt
+#'                                  models = models,
+#'                                  samples = 10,
+#'                                  horizon = 7,
+#'                                  weighting_period = 5)
+#'
+#'
+#' # together with forecast_rt
+#' fc_rt <- forecast_rt(EpiSoon::example_obs_rts[1:10, ],
+#'                      model = function(...){
+#'                        stackr_mixture_model(models = models,
+#'                                             weighting_period = 5,
+#'                                             ...)},
+#'                      samples = 10,
+#'                      horizon = 7)
+#'
+#' forecast_eval <- evaluate_model(EpiSoon::example_obs_rts,
+#'                                 EpiSoon::example_obs_cases,
+#'                                 model = function(...){
+#'                                   stackr_mixture_model(models = models,
+#'                                                        weighting_period = 5,
+#'                                                        ...)},
+#'                                 horizon = 7, samples = 10,
+#'                                 serial_interval = example_serial_interval,
+#'                                min_points = 10)
+#'
+#' plot_forecast_evaluation(forecast_eval$forecast_rts,
+#'                          EpiSoon::example_obs_rts,
+#'                          horizon_to_plot = 7)
+#'
+#' }
+#'
+
+stackr_mixture_model <- function(y = NULL,
+                                 models = NULL,
+                                 samples = NULL,
+                                 horizon = NULL,
+                                 weighting_period = 5) {
+
+
+  #### Error Handling
+  # check if y is there
+  if(is.null(y)) stop("parameter y is missing")
+
+
+  # check if stackr is installed
+  check_suggests("stackr",
+                 dev_message = "Install using devtools::install_github('nikosbosse/stackr')")
+
+  if (length(y) <= weighting_period) {
+    stop("not enough observations to do weighting. Adjust weighting_period")
+  }
+
+  #### split data into train data and data for weighting. Use train data to
+  # generate forecasts, score them against weight data, then generate forecasts
+  # bases on the entire time series and use the obtained weights to create mixture
+  n <- length(y)
+  y_train <- y[1:(n - weighting_period)]
+  y_weight <- y[(n - weighting_period + 1):n]
+
+  #### fit models on train data and generate forecasts
+  fc_w <- purrr::map_dfr(seq_along(models),
+                         .f = function(i) {
+                           model_function <- models[[i]]
```

Comment:
can just do models[[i]]() directly

Reply:
don't think so. models[i] is still a list.

Reply:
`models[[]]()`

Reply:
alternatively and better pass the list of models in and make i -> model then you can just call model()

Reply:
I use the i though when I call names(models)[i]
not sure what you mean with this `models[[]]()`
happy to change but not clear what to do

Reply:
aah. now saw the comment about double brackets.

Reply:
left the i unchanged for now. again happy to change if you have a better suggestion